File: src/platform/llm/implementations/openai.ts
import { LLMProvider, GenerateRequest, GenerateResponse, LLMProviderCapability } from '../types';
import { ProviderError, RateLimitError, AuthenticationError, InvalidRequestError } from '../../../core/errors/llm-errors';
import { Logger } from '../../../core/telemetry';

interface OpenAIMessage {
  role: 'system' | 'user' | 'assistant';
  content: string;
}

interface OpenAIRequest {
  model: string;
  messages: OpenAIMessage[];
  temperature?: number;
  max_tokens?: number;
  stop?: string[];
  response_format?: { type: 'json_object' | 'text' };
}

interface OpenAIResponse {
  choices?: Array<{
    message?: { content?: string };
    finish_reason?: string;
  }>;
  usage?: {
    prompt_tokens: number;
    completion_tokens: number;
    total_tokens: number;
  };
}

interface OpenAIErrorResponse {
  error?: {
    message: string;
    type?: string;
    code?: string;
  };
}

export class OpenAIProvider implements LLMProvider {
  readonly providerId = 'openai';
  readonly capabilities: LLMProviderCapability = {
    capabilities: new Set(['json_mode', 'streaming', 'function_calling', 'vision']),
    maxTokens: 4096,
    contextWindow: 128000,
  };

  constructor(private config: { apiKey: string; model: string }) {}

  async generate(request: GenerateRequest): Promise<GenerateResponse> {
    try {
      if (!this.config.apiKey) throw new Error('Missing API Key');

      const messages: OpenAIMessage[] = [];
      if (request.systemPrompt) {
        messages.push({ role: 'system', content: request.systemPrompt });
      }
      messages.push({ role: 'user', content: request.prompt });

      const body: OpenAIRequest = {
        model: this.config.model,
        messages,
        temperature: request.temperature ?? 0.7,
        max_tokens: request.maxTokens ?? this.capabilities.maxTokens,
      };

      if (request.jsonMode) {
        body.response_format = { type: 'json_object' };
      }

      if (request.stopSequences && request.stopSequences.length > 0) {
        body.stop = request.stopSequences;
      }

      Logger.debugProvider(this.providerId, 'REQUEST', body);

      const response = await fetch('https://api.openai.com/v1/chat/completions', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${this.config.apiKey}`,
        },
        body: JSON.stringify(body),
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({})) as OpenAIErrorResponse;
        const errorMessage = errorData.error?.message || response.statusText;
        throw new Error(`${response.status}: ${errorMessage}`);
      }

      const data: OpenAIResponse = await response.json();
      Logger.debugProvider(this.providerId, 'RESPONSE', data);

      const choice = data.choices?.[0];
      const text = choice?.message?.content?.trim() || '';
      const usage = data.usage || { prompt_tokens: 0, completion_tokens: 0, total_tokens: 0 };

      return {
        text,
        usage: {
          promptTokens: usage.prompt_tokens,
          completionTokens: usage.completion_tokens,
          totalTokens: usage.total_tokens,
        },
        finishReason: this.mapFinishReason(choice?.finish_reason ?? 'unknown'),
      };
    } catch (e: unknown) {
      throw this.normalizeError(e);
    }
  }

  private mapFinishReason(reason: string): GenerateResponse['finishReason'] {
    switch (reason) {
      case 'stop': return 'stop';
      case 'length': return 'length';
      case 'content_filter': return 'content_filter';
      default: return 'unknown';
    }
  }

  private normalizeError(e: unknown): Error {
    const message = e instanceof Error ? e.message : String(e);
    if (message.includes('429')) return new RateLimitError(message);
    if (message.includes('401') || message.includes('403')) return new AuthenticationError(message);
    if (message.includes('400')) return new InvalidRequestError(message);
    return new ProviderError(message);
  }
}

File: crates/wfs-core/src/llm/openai.rs
use async_trait::async_trait;
use reqwest::Client;
use std::collections::HashSet;

use crate::error::LlmError;
use crate::types::{FinishReason, GenerateRequest, GenerateResponse, ModelCapability, ProviderCapabilities, Usage};
use super::LlmProvider;

pub struct OpenAIProvider {
    api_key: String,
    model: String,
    client: Client,
    caps: ProviderCapabilities,
}

impl OpenAIProvider {
    pub fn new(api_key: String, model: String) -> Self {
        Self {
            api_key,
            model,
            client: Client::new(),
            caps: ProviderCapabilities {
                capabilities: HashSet::from([
                    ModelCapability::JsonMode,
                    ModelCapability::Streaming,
                    ModelCapability::FunctionCalling,
                    ModelCapability::Vision,
                ]),
                max_tokens: 4096,
                context_window: 128000,
            },
        }
    }
}

#[async_trait]
impl LlmProvider for OpenAIProvider {
    fn provider_id(&self) -> &'static str {
        "openai"
    }

    fn capabilities(&self) -> &ProviderCapabilities {
        &self.caps
    }

    async fn generate(&self, request: &GenerateRequest) -> Result<GenerateResponse, LlmError> {
        let mut messages = Vec::new();
        if let Some(ref system) = request.system_prompt {
            messages.push(serde_json::json!({"role": "system", "content": system}));
        }
        messages.push(serde_json::json!({"role": "user", "content": &request.prompt}));

        let mut body = serde_json::json!({
            "model": self.model,
            "messages": messages,
            "temperature": request.temperature.unwrap_or(0.7),
            "max_tokens": request.max_tokens.unwrap_or(self.caps.max_tokens),
        });

        if request.json_mode.unwrap_or(false) {
            body["response_format"] = serde_json::json!({"type": "json_object"});
        }

        if let Some(ref stop) = request.stop_sequences {
            if !stop.is_empty() {
                body["stop"] = serde_json::json!(stop);
            }
        }

        let response = self
            .client
            .post("https://api.openai.com/v1/chat/completions")
            .header("Authorization", format!("Bearer {}", self.api_key))
            .json(&body)
            .send()
            .await
            .map_err(|e| normalize_error(&format!("Request failed: {e}")))?;

        let status = response.status();
        if !status.is_success() {
            let error_text = response.text().await.unwrap_or_default();
            return Err(normalize_error(&format!("{status}: {error_text}")));
        }

        let data: serde_json::Value = response
            .json()
            .await
            .map_err(|e| LlmError::Parse(format!("Failed to parse response: {e}")))?;

        let choice = data["choices"][0].clone();
        let text = choice["message"]["content"]
            .as_str()
            .unwrap_or("")
            .trim()
            .to_string();

        let usage = Usage {
            prompt_tokens: data["usage"]["prompt_tokens"].as_u64().unwrap_or(0) as u32,
            completion_tokens: data["usage"]["completion_tokens"].as_u64().unwrap_or(0) as u32,
            total_tokens: data["usage"]["total_tokens"].as_u64().unwrap_or(0) as u32,
        };

        let finish_reason = match choice["finish_reason"].as_str() {
            Some("stop") => FinishReason::Stop,
            Some("length") => FinishReason::Length,
            Some("content_filter") => FinishReason::ContentFilter,
            _ => FinishReason::Unknown,
        };

        Ok(GenerateResponse { text, usage, finish_reason })
    }
}

fn normalize_error(msg: &str) -> LlmError {
    if msg.contains("429") || msg.to_lowercase().contains("rate limit") {
        LlmError::RateLimit(msg.to_string())
    } else if msg.contains("401") || msg.contains("403") {
        LlmError::Authentication(msg.to_string())
    } else if msg.contains("400") || msg.contains("422") {
        LlmError::InvalidRequest(msg.to_string())
    } else if msg.contains("timeout") || msg.to_lowercase().contains("timed out") {
        LlmError::Timeout(msg.to_string())
    } else {
        LlmError::Provider(msg.to_string())
    }
}
